cloud-controller-manager/do/firewall_controller.go

```diff
@@ -290,12 +369,54 @@ func isEqual(targetFirewall *godo.Firewall, fr *godo.FirewallRequest) (bool, []s
 }
 
 func (fm *firewallManager) createFirewall(ctx context.Context, fr *godo.FirewallRequest) (*godo.Firewall, error) {
-	currentFirewall, _, err := fm.client.Firewalls.Create(ctx, fr)
+	currentFirewall, err := func() (*godo.Firewall, error) {
+		var (
+			code   int
+			method string
+		)
+		// The ObserverFunc gets called by the deferred ObserveDuration. The
+		// method and code values will be set before ObserveDuration is called
+		// with the value returned from the response from the Firewall API request.
+		timer := prometheus.NewTimer(prometheus.ObserverFunc(func(v float64) {
+			fm.metrics.apiRequestDuration.With(prometheus.Labels{"method": method, "code": strconv.FormatInt(int64(code), 10)}).Observe(v)
+		}))
+		defer timer.ObserveDuration()
+		fw, resp, err := fm.client.Firewalls.Create(ctx, fr)
+		if resp != nil {
+			code = resp.StatusCode
+			if resp.Request != nil {
+				method = resp.Request.Method
+			}
+		}
+		return fw, err
+	}()
+
 	return currentFirewall, err
 }
 
 func (fm *firewallManager) updateFirewall(ctx context.Context, fwID string, fr *godo.FirewallRequest) (*godo.Firewall, error) {
-	currentFirewall, _, err := fm.client.Firewalls.Update(ctx, fwID, fr)
+	currentFirewall, err := func() (*godo.Firewall, error) {
```

Comment:
nit: I think the reason you're using a callback approach in `firewallManager.Get` is that more is happening in that function than simply calling out to the firewalls API whereas here in `firewallManager.updateFirewall` we are simply wrapping a single call to the firewalls API; removing the callback would lead to a somewhat cleaner implementation here.

Reply:
nice catch!